### PR TITLE
Make ZoneGridModel more forgiving (handle possibly out of sync state)

### PR DESCRIPTION
**Pull request by febbraiod**
… state)

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: cmp/zoneGrid/ZoneGridModel.ts [hunk 1]
```diff
@@ -616,41 +625,65 @@ export class ZoneGridModel extends HoistModel {
...
+                            );
+                            offenders.push(it.field);
+                        }
                     });
+
+                    if (!isEmpty(offenders)) {
+                        remove(mapping, it => offenders.includes(it.field));
+                    }
+
+                    if ((zone == 'tl' || zone == 'tr') && isEmpty(mapping)) {
```
Comment (TomTirapani) [inline cmp/zoneGrid/ZoneGridModel.ts:675]: Great catch! But I think we want to do this outside of the `!isEmpty(limit.only)` condition.

(I know enforcing `only` is currently the only place where we remove mappings, but I think the intent is clearer if we always do this as the final check of parseMappings)

**Review by TomTirapani — COMMENTED** [review]
Thanks for catching and taking a look at this issues Don!

It's certainly clearly problematic if this thing throws because saved state is no longer valid! I also like how you are resolving back to the default state to get back to a good state 👍  But I realised that assumes the **default** state respects the constraints, and `parseMappings()` is used both to set up the initial default state **and** to handle changing the state.

I wonder if we should only be this lenient in the latter case? It seems wrong to allow apps to set up a Grid where the defaults are always wrong - especially if we expect loading the defaults to resolve the issue. Maybe we could add a second `strict` parameter to `parseMappings()`, and conditionally throw or warn based on that?

File: cmp/zoneGrid/ZoneGridModel.ts (hunk 1 above)
Comment (febbraiod) [reply cmp/zoneGrid/ZoneGridModel.ts:675]: Great note Tom. I considered the same, but apparently made the wrong choice ;)

File: cmp/zoneGrid/ZoneGridModel.ts [hunk 2]
```diff
@@ -635,52 +637,58 @@ export class ZoneGridModel extends HoistModel {
                 if (col) {
                     mapping.push(ret);
                 } else {
-                    console.warn(`Column not found for field '${ret.field}'`);
+                    const message = `Column not found for field '${ret.field}'`;
+                    this.throwOrWarn(message, strict);
```
Comment (TomTirapani) [inline cmp/zoneGrid/ZoneGridModel.ts:641]: We could also do something like this:
```
const throwOrWarn = strict ? throwIf : warnIf;
throwOrWarn(`Column not found for field '${ret.field}'`);
```

File: cmp/zoneGrid/ZoneGridModel.ts [hunk 3]
```diff
@@ -611,46 +621,77 @@ export class ZoneGridModel extends HoistModel {
...
                 const ret = isString(it) ? {field: it} : it,
                     col = this.findColumnSpec(ret);
 
-                throwIf(!col, `Column not found for field ${ret.field}`);
-                return mapping.push(ret);
+                if (col) {
+                    mapping.push(ret);
+                } else {
+                    const message = `Column not found for field '${ret.field}'`;
+                    this.throwOrWarn(message, strict);
```
Comment (febbraiod) [inline cmp/zoneGrid/ZoneGridModel.ts:641]: inline function that captures strict.

File: cmp/zoneGrid/ZoneGridModel.ts [hunk 4]
```diff
@@ -611,46 +621,77 @@ export class ZoneGridModel extends HoistModel {
...
+                    mapping = this._defaultState.mappings[zone];
+                }
+
+                if (isFinite(limit.max) && mapping.length > limit.max) {
+                    const diff = mapping.length - limit.max,
+                        message = `Exceeds maximum ${limit.max} mappings in zone "${zone}". Dropping last ${diff} fields`;
+
+                    this.throwOrWarn(message, strict);
+
+                    mapping = dropRight(mapping, mapping.length - limit.max);
```
Comment (febbraiod) [inline cmp/zoneGrid/ZoneGridModel.ts:662]: use diff var.

Comment (febbraiod) [reply cmp/zoneGrid/ZoneGridModel.ts:662]: set to default here as well

File: cmp/zoneGrid/ZoneGridModel.ts [hunk 5]
```diff
@@ -611,46 +621,77 @@ export class ZoneGridModel extends HoistModel {
...
+                            const message = `Field "${it.field}" not allowed in zone "${zone}. Removing"`;
+
+                            this.throwOrWarn(message, strict);
+
+                            offenders.push(it.field);
+                        }
                     });
+
+                    if (!isEmpty(offenders)) {
+                        remove(mapping, it => offenders.includes(it.field));
```
Comment (febbraiod) [inline cmp/zoneGrid/ZoneGridModel.ts:678]: set to default again here. Only need to warn for one field and say we're going to the default immediately at that point.

File: cmp/zoneGrid/ZoneGridModel.ts (hunk 3 above)
Comment (febbraiod) [reply cmp/zoneGrid/ZoneGridModel.ts:641]: this can also do the reset to default. handleConstraint.

**Pull request merged**